packages/rich-text-links/src/index.ts

```diff
@@ -1,80 +1,56 @@
 export type EntityLink = SysObject['sys'];
 export type SysObject = Link<EntityType>;
-
-type LinkCache = Cache;
+export type EntityLinkNodeData = {
+  target: SysObject;
+};
 
 /**
- * Implements a subset of the Set API:
- * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
+ * Given a rich text document, returns all entity links.
  */
-class Cache {
-  cache: Record<string, true>;
-
-  constructor() {
-    this.cache = {};
-  }
-
-  add(id: string): void {
-    this.cache[id] = true;
+export function getRichTextEntityLinks(document: Document): EntityLinks {
+  const entityLinks: EntityLinkMaps = {
+    Entry: new Map(),
+    Asset: new Map(),
+  };
+
+  const content = (document && document.content) || ([] as Node[]);
+  for (const node of content) {
+    addLinksFromRichTextNode(node, entityLinks);
   }
 
-  has(id: string): Boolean {
-    return this.cache[id] || false;
-  }
+  return {
+    Entry: Array.from(entityLinks.Entry.values()),
+    Asset: Array.from(entityLinks.Asset.values()),
+  };
 }
 
-/**
- * Given a rich text document and an (optional) entity type, returns all unique
- * matching entity links.
- *
- * If there is no linkType provided, returns all unique entity links.
- */
-export function getRichTextEntityLinks(document: Document, linkType?: EntityType): EntityLink[] {
-  const links: EntityLink[] = [];
-  const linkCache: LinkCache = new Cache();
-
-  for (const node of document.content) {
-    addLinksFromRichTextNode(links, node, linkCache, linkType);
-  }
-
-  return links;
-}
-
-function addLinksFromRichTextNode(
-  links: EntityLink[],
-  node: Node,
-  linkCache: LinkCache,
-  linkType?: EntityType,
-): EntityLink[] {
+function addLinksFromRichTextNode(node: Node, links: EntityLinkMaps): void {
   const toCrawl: Node[] = [node];
 
   while (toCrawl.length > 0) {
     const { data, content } = toCrawl.pop() as Block;
 
-    if (isMatchingLinkObject(data, linkType) && !linkCache.has(data.target.sys.id)) {
-      links.push(data.target.sys);
-      linkCache.add(data.target.sys.id);
+    if (isLinkObject(data)) {
+      links[data.target.sys.linkType].set(data.target.sys.id, data.target.sys);
     } else if (Array.isArray(content)) {
       for (const childNode of content) {
         toCrawl.push(childNode);
       }
     }
   }
-
-  return links;
 }
 
-function isMatchingLinkObject(data: NodeData, linkType?: EntityType): Boolean {
+function isLinkObject(data: NodeData): data is EntityLinkNodeData {
   const sys = data && (data.target as SysObject) && (data.target.sys as EntityLink);
-  const isLinkObject = sys && typeof sys.id === 'string' && sys.type === 'Link';
-  if (!isLinkObject) {
-    return false;
-  }
-  if (linkType) {
-    return sys.linkType === linkType;
-  }
-  return sys.linkType === 'Entry' || sys.linkType === 'Asset';
+  return (
+    sys &&
+    typeof sys.id === 'string' &&
+    sys.type === 'Link' &&
+    (sys.linkType === 'Entry' || sys.linkType === 'Asset')
+  );
```

Comment:
I'd prefer keeping the previous version, as it's more readable what constraints are checked to determine whether it's a link object or not.

```typescript
const isLinkObject = sys && typeof sys.id === 'string' && sys.type === 'Link';

if (!isLinkObject) {
  return false;
}

const isKnownLinkType = sys.linkType === 'Entry' || sys.linkType === 'Asset';

return isKnownLinkType
```

But that's just a personal preference of course 🙂

Reply:
👍